walkthrough/05-aws/04-reusing-stacks/README.md

```diff
@@ -0,0 +1,179 @@
+
+This is a limitation of the design of RFC [#3313](https://github.com/gruntwork-io/terragrunt/issues/3313), and is something that additional tooling might be added in the future to make this more convenient to handle.
+
+The design philosophy of the proposal is that the authors Unit `terragrunt.hcl` files should decide how they want to handle these sorts of conflicts, as they are authoring the configurations that will be passed directly as inputs to OpenTofu/Terraform modules.
+
+In this case, the `path` gives us a useful way of disambiguating the resources that are created by the `api` Unit. We can update the `name` input to reference the name of the directory it is placed in.
+
+```hcl
+...
+	name = "${local.project_name}-${basename(get_terragrunt_dir())}-${local.random_id}-${local.environment}"
+...
+```
+
+Now, if you were to re-run the `mock-stack-generate.sh` script, and apply the Stacks, you would see that the resources are created without any issues.
+
+```bash
+$ ./mock-stack-generate.sh
+$ terragrunt run-all apply --terragrunt-non-interactive
+```
+
+We'll now have resources provisioned like `stacks-walkthrough-05-04-api-zgggh-dev` and `stacks-walkthrough-05-04-canary-api-zgggh-dev`.
+
+To drive the point home, let's add one more, and see that the logic works as expected:
+
+```hcl
+unit "api" {
+	source = "../../../../units/api"
+	path   = "services/api"
+}
+
+unit "canary_api" {
+    source = "../../../../units/api"
+    path   = "services/canary-api"
+}
+
+unit "backup_api" {
+    source = "../../../../units/api"
+    path   = "services/backup-api"
+}
+
+unit "db" {
+	source = "../../../../units/db"
+	path   = "storage/db"
+}
+```
+
+```bash
+$ ./mock-stack-generate.sh
+$ terragrunt run-all apply --terragrunt-non-interactive
+```
+
+Note that both the `dev` and `prod` environments continuously received these updates simultaneously. This doesn't have to be the case.
+
+The references here could have been updated to reference versioned directories (e.g. `services/api/v1`), and the `terragrunt.stack.hcl` file in each environment could be updated in a rolling fashion to adopt the new version of the Stack. The reference could also be a tagged reference to a different repository (e.g. `github.com/yhakbar/terragrunt-3313-stacks-walkthrough//walkthrough/05-aws/04-reusing-stacks?ref=v1.0.0`).
+
+Also note that this mechanism scales for without any ambiguity as to where `inputs` are being defined. They are always defined for a Unit on the `terragrunt.hcl` file itself. No matter how deeply nested a Stack definition is (a Stack defining a Stack defining a Stack, for example), the `inputs` on a `terragrunt.hcl` file always determines the inputs for the Unit.
+
+## Cleanup
+
+To clean up the resources that were provisioned in this chapter, run the following:
+
+```bash
+$ terragrunt run-all destroy --terragrunt-non-interactive
+```
+
+## Next Steps
+
+> TBD
+>
+> Please submit feedback if you would like to see more content here.
```

Comment:
One thing missing here is the canonical use case of specifying different configurations for different stack instances using plain old literals. Could you show how each stack would receive different literal values, similar to how each `terragrun.hcl` file can specify literals directly in the `inputs`?

Reply:
They don't.

The design here is that you always define all the configurations for all of your inputs to OpenTofu/Terraform modules in the `terragrunt.hcl` files. Full stop. There's no hidden secondary way to override values.

Authors of `terragrunt.hcl` files have to decide that they want to _pull_ external configurations like the [environment.hcl file](https://github.com/yhakbar/terragrunt-3313-stacks-walkthrough/blob/main/walkthrough/05-aws/03-no-shared/live/dev/environment.hcl) instead of having overriding configurations _pushed_ from above further removed from the context of how the inputs are going to be used in OpenTofu/Terraform modules.

Reply:
I mentioned above that I tried to tackle that here:
https://github.com/yhakbar/terragrunt-3313-stacks-walkthrough/tree/main/walkthrough/05-aws/04-reusing-stacks#stack-dynamicity